src/OpenTelemetry/Metrics/MetricExporter.cs

```diff
@@ -0,0 +1,96 @@
+// <copyright file="MetricExporter.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+
+namespace OpenTelemetry.Metrics
+{
+    public abstract class MetricExporter : IDisposable
+    {
+        private int shutdownCount;
+
+        public BaseProvider ParentProvider { get; internal set; }
+
+        public abstract ExportResult Export(IEnumerable<Metric> metrics);
+
+        public bool Shutdown(int timeoutMilliseconds = Timeout.Infinite)
+        {
+            if (timeoutMilliseconds < 0 && timeoutMilliseconds != Timeout.Infinite)
+            {
+                throw new ArgumentOutOfRangeException(nameof(timeoutMilliseconds), timeoutMilliseconds, "timeoutMilliseconds should be non-negative.");
+            }
+
+            if (Interlocked.Increment(ref this.shutdownCount) > 1)
+            {
+                return false; // shutdown already called
+            }
+
+            try
+            {
+                return this.OnShutdown(timeoutMilliseconds);
+            }
+            catch (Exception)
+            {
+                return false;
+            }
+        }
+
+        public bool ForceFlush(int timeoutMilliseconds = Timeout.Infinite)
+        {
+            if (timeoutMilliseconds < 0 && timeoutMilliseconds != Timeout.Infinite)
+            {
+                throw new ArgumentOutOfRangeException(nameof(timeoutMilliseconds), timeoutMilliseconds, "timeoutMilliseconds should be non-negative.");
+            }
+
+            try
+            {
+                return this.OnForceFlush(timeoutMilliseconds);
+            }
+            catch (Exception)
+            {
+                return false;
+            }
+        }
+
+        public virtual AggregationTemporality GetAggregationTemporality()
+        {
+            return AggregationTemporality.Cumulative;
```

Comment:
Do we want this to be `AggregationTemporality.Unspecified`?

Reply:
For exporter, I would suggest have two GET properties `MetricExporter.SupportedAggregationTemporality` and `MetricExporter.PreferredAggregationTemporality`.

```csharp
[Flags]
public enum AggregationTemporality : byte
{
    Unspecified = 0x00,
    Cumulative = 0x01,
    Delta = 0x02,
    Both = 0x03,
}
```

| Exporter   | SupportedAggregationTemporality | PreferredAggregationTemporality  |
| ----------- | ----------- | ---- |
| Console | Both | Cumulative |
| InMemory | Both | Unspecified |
| OTLP | Both | Unspecified |
| Prometheus | Cumulative | Cumulative |
| Statsd | Delta | Delta |

Reply:
I had thought that Cumulative and Delta were required for Prometheus and Statsd, respectively. If my assumption is right then what's the purpose of `PreferredAggregationTemporality`?

Reply:
For Prometheus/StatsD, both would be same. "Preferred" is for those exporters capable of handling both.

Reply:
Here is my thinking:
* There will be exporters that:
  1. only support Delta, so we have to either convert Cumulative to Delta, or drop the data, or do something else - e.g. treat it as Gauge 2
  2. only support Cumulative, so we have to either convert Delta to Cumulative, or drop the data
  3. supports both Cumulative and Delta (e.g. Console and InMemory exporter), but would prefer a certain way by default, and allow the user to change the behavior